src/ppx/metrics.cpp

```diff
@@ -0,0 +1,219 @@
+    const size_t percentileIndex99 = entriesCount * 99 / 100;
+    statistics.percentile99        = sorted[percentileIndex99].value;
+
+    return statistics;
+}
+
+void MetricGauge::UpdateBasicStatistics(double seconds, double value)
+{
+    mAccumulatedValue += value;
+
+    mBasicStatistics.min = std::min(mBasicStatistics.min, value);
+    mBasicStatistics.max = std::max(mBasicStatistics.max, value);
+
+    const size_t entriesCount = mTimeSeries.size();
+    mBasicStatistics.average  = mAccumulatedValue / entriesCount;
+    if (entriesCount > 1) {
+        const double timeSpan      = mTimeSeries.back().seconds - mTimeSeries.front().seconds;
+        mBasicStatistics.timeRatio = mAccumulatedValue / timeSpan;
+    }
+    else {
+        mBasicStatistics.timeRatio = mTimeSeries.front().value;
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+MetricCounter::MetricCounter(const MetricMetadata& metadata)
+    : mMetadata(metadata), mCounter(0U)
+{
+}
+
+MetricCounter::~MetricCounter()
+{
+}
+
+uint64_t MetricCounter::Increment(uint64_t add)
+{
+    mCounter += add;
+    return mCounter;
+}
+
+uint64_t MetricCounter::Get() const
+{
+    return mCounter;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Run::Run(const char* name)
+    : mName(name)
+{
+    PPX_ASSERT_MSG(name != nullptr, "A run name cannot be null");
+}
+
+Run::~Run()
+{
+    for (auto [name, metric] : mGauges) {
+        delete metric;
+    }
+    mGauges.clear();
+    for (auto [name, metric] : mCounters) {
+        delete metric;
+    }
+    mCounters.clear();
+}
+
+void Run::AddMetric(MetricGauge* metric)
+{
+    PPX_ASSERT_NULL_ARG(metric);
+    const auto ret = mGauges.insert({metric->mMetadata.name, metric});
+    PPX_ASSERT_MSG(ret.second, "An insertion shall always take place when adding a metric");
+}
+
+void Run::AddMetric(MetricCounter* metric)
+{
+    PPX_ASSERT_NULL_ARG(metric);
+    const auto ret = mCounters.insert({metric->mMetadata.name, metric});
+    PPX_ASSERT_MSG(ret.second, "An insertion shall always take place when adding a metric");
+}
+
+bool Run::HasMetric(const char* name) const
+{
+    return mGauges.find(name) != mGauges.end() || mCounters.find(name) != mCounters.end();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Manager::Manager()
+{
+}
+
+Manager::~Manager()
+{
+    for (auto [name, run] : mRuns) {
+        delete run;
+    }
+    mRuns.clear();
+}
+
+Run* Manager::AddRun(const char* name)
+{
+    PPX_ASSERT_MSG(name != nullptr, "A run name must not be null");
+    PPX_ASSERT_MSG(mRuns.find(name) == mRuns.end(), "Runs must have unique names (duplicate name detected)");
+
+    Run* run = new Run(name);
+    if (run == nullptr) {
+        return nullptr;
+    }
+
+    const auto ret = mRuns.insert({name, run});
+    PPX_ASSERT_MSG(ret.second, "An insertion shall always take place when adding a run");
```

Comment:
Does this causes a "unused variable ret" warning in release builds?

Reply:
Good point, but no. It seems the BW assert macro always uses the condition to do some logging. Whether the assert triggers or not is all from cassert.

Reply:
Thanks for checking :)